### PR TITLE
Refactor containerisation 2

### PR DESCRIPTION
**Pull request by godrei**
### Checklist

- [x] I've read and followed the [Contribution Guidelines](https://github.com/bitrise-io/bitrise/blob/master/.github/CONTRIBUTING.md)
- [x] `README.md` is updated with the changes (if needed)

### Version

Requires a *MAJOR/MINOR/PATCH* [version update](https://semver.org/)

### Context

Containerization-related configuration is getting refactored:

1, Moving service and container configuration to new top-level entities (services and containers): https://github.com/bitrise-io/bitrise/pull/952

2, Step grouping mechanism with a new step list item (`with`)

From this:

```
workflows:
  _setup:
    steps:
    - activate-ssh-key: { }
    - git-clone: {}

  test:
    before_run:
    - _setup
    container: ruby
    services:
    - postgres
    steps:
    - script:
        title: Setup DB
        inputs:
        - content: bundle exec rails db:setup
    - script:
        title: Run tests
        inputs:
        - content: bundle exec rspec
```

to this:

```
workflows:
  test:
    steps:
    - activate-ssh-key@4: { }
    - git-clone@8: {}
    - with:
        container: ruby
        services:
        - postgres
        steps:
        - script:
            title: Setup DB
            inputs:
            - content: bundle exec rails db:setup
        - script:
            title: Run tests
            inputs:
            - content: bundle exec rspec
```

The two changes will be introduced in separate PRs to the `refactor-containerisation` branch.

This PR implements the second change: Step grouping mechanism with a new step list item (`with`)

### Changes

bitrise.yml model changes:
- New `with` step list item under `<workflow_id>.steps`, which allows setting step execution and service containers for a group of steps
- Step execution and service containers removed from the workflows

Workflow run changes:
- The list of steps to be executed within the workflow is now controlled by the extended WorkflowRunPlan, rather than the actual workflow
- WorkflowRunPlan contains a homogenous list of steps, for steps within a `with` group the required container and services along the group ID are stored in the StepExecutionPlan
- These new StepExecutionPlan properties control the lifecycle of containers through the execution of the given step list

Docker container manager changes:
- The lifecycle of containers is tight to the lifecycle of `with` step groups, the with group's uuid is used to reference the related running containers
- The container name has changes: `bitrise-workflow-<workflow_id>` -> `bitrise-workflow-<group_uuid>`

### CODE REVIEW

File: cli/run.go [hunk 1]
```diff
@@ -166,6 +166,15 @@ func setupAgentConfig() (*configs.AgentConfig, error) {
 	return &config, nil
 }
 
+type DockerManager interface {
```
Comment (godrei) [inline cli/run.go:169]: Moved to this file from run_util.go, because it gets used in this file only.

File: cli/run.go [hunk 2]
```diff
@@ -322,10 +333,8 @@ func (r WorkflowRunner) runWorkflows(tracker analytics.Tracker) (models.BuildRun
 	for i, workflowRunPlan := range plan.ExecutionPlan {
 		isLastWorkflow := i == len(plan.ExecutionPlan)-1
 		workflowToRun := r.config.Config.Workflows[workflowRunPlan.WorkflowID]
-		if workflowToRun.Title == "" {
```
Comment (godrei) [inline cli/run.go:325]: Determining the workflow title is moved to the `createWorkflowRunPlan` function.

File: models/models.go [hunk 3]
```diff
@@ -11,11 +11,25 @@ import (
 
 const (
 	// FormatVersion ...
-	FormatVersion = "14"
+	FormatVersion       = "14"
+	StepListItemWithKey = "with"
 )
 
+// WithModel ...
```
Comment (ofalvai) [inline models/models.go:18]: nit: these useless comments are no longer required by the linter, you don't need to keep writing them. I also like to clean up existing ones if I touch a file.

File: cli/run_util.go [hunk 4]
```diff
@@ -566,103 +566,126 @@ func (r WorkflowRunner) runStep(
...
+	if err != nil {
+		log.Debugf("Couldn't read envs from envman.")
+	}
+
+	if containerID != "" {
+		containerDef := r.ContainerDefinition(containerID)
+		if containerDef != nil {
+			log.Infof("ℹ️ Running workflow in docker container: %s", containerDef.Image)
+
+			// TODO: Why only login for step execution containers?
```
Comment (ofalvai) [inline cli/run_util.go:591]: I think the original assumption was that services containers are public utility images like `postgres`. But it should also support authorization I think

File: cli/run_util.go [hunk 5]
```diff
@@ -566,103 +566,126 @@ func (r WorkflowRunner) runStep(
...
+		servicesDefs := r.ServiceDefinitions(serviceIDs...)
+		_, err := r.dockerManager.StartServiceContainersFroStepGroup(servicesDefs, groupID, envList)
+		if err != nil {
+			log.Errorf("❌ Some services failed to start properly!")
+		}
+	}
+}
+
+func (r WorkflowRunner) stopContainersForStepGroup(groupID, workflowTitle string) {
+	if container := r.dockerManager.GetContainerFroStepGroup(groupID); container != nil {
```
Comment (ofalvai) [inline cli/run_util.go:613]: nit: typo

```suggestion
	if container := r.dockerManager.GetContainerForStepGroup(groupID); container != nil {
```

**Review by ofalvai — APPROVED** [review]

File: models/models.go (hunk 3 above)
Comment (godrei) [reply models/models.go]: Cleaned up all the modified files in https://github.com/bitrise-io/bitrise/pull/956/commits/25e96f7785ddff7aa77bdb34498c5711242e2232

**Review by ofalvai — APPROVED** [review]

**Pull request merged**